### ISSUE TITLE
Allow Minecraft-version-specific subfolders in the mods folder

### ISSUE THREAD

**Issue by Earthcomputer** [QuiltMC/quilt-loader#15]
The main use case is for people who use the vanilla launcher, which only has one mods folder shared between potentially many instances. Forge allows you to create subfolders in the mods folder named the same as each Minecraft version, and FML will not attempt to load mods from subfolders not matching your current version. I know quite a few people find it annoying that Fabric loader does not allow you to do the same thing, so it would be nice if Quilt loader had this feature.

**Comment by TwoLeggedCat** [comment]
Can't you use multiple different installations with separate game directories for different versions? We wouldn't want to encourage people to use the same game directory across several versions anyway.

**Comment by williambl** [comment]
IMO FML's doing this is a relic from before it was easy to separate instances into their own directories.

Perhaps we should instead make quilt installer optionally create a new directory for the quilt install?

**Comment by Geolykt** [comment]
> Can't you use multiple different installations with separate game directories for different versions? We wouldn't want to encourage people to use the same game directory across several versions anyway.

It's complicated and I recall mods not working nicely with that; though that experience must be a few years old by now.

**Comment by Siuolthepic** [comment]
> > Can't you use multiple different installations with separate game directories for different versions? We wouldn't want to encourage people to use the same game directory across several versions anyway.
> 
> It's complicated and I recall mods not working nicely with that; though that experience must be a few years old by now.

I have had no complications with such a thing. On Vanilla launcher its as simple as changing the installation folder to somewhere else and then dragging the mods in there instead.

**Comment by chexo3** [comment]
I see no reason not to. It's simple to add and makes things easier for less technical users. We seem to forget not everyone uses MultiMC (even though they arguably should)

**Comment by TwoLeggedCat** [comment]
> We seem to forget not everyone uses MultiMC (even though they arguably should)

What does this have to do with MultiMC? This is possible even with Mojang's launcher. Doing this in Quilt would just encourage people to use the same game directory across multiple versions *and* with different mods, often leadding to corruption worlds, and offer no real benefit over what the launcher itself can do.

**Comment by hYdos** [comment]
>  Additionally, I'm not sure you have to be a "technical user" to put different versions in different folders, and no more so than putting different mods in the correct folders.

have you seen fabricoord's #player-support ?

**Comment by TwoLeggedCat** [comment]
> > Additionally, I'm not sure you have to be a "technical user" to put different versions in different folders, and no more so than putting different mods in the correct folders.
> 
> have you seen fabricoord's #player-support ?

Yes, and I have no more confidence in those who cannot use a separate profile to use a separate folder for different versions than a separate profile.

> > often leading to corrupted worlds
> 
> As long as you don't join the world with incompatible modsets that is not possible. And I think there should be good enough guards at warning the user if they attempted to do that

This was more directed at the idea of running different version in the same folder than it was different mods.

> IMO FML's doing this is a relic from before it was easy to separate instances into their own directories.
> 
> Perhaps we should instead make quilt installer optionally create a new directory for the quilt install?

Assuming by "quilt install" you mean the profile the installer can create, this seems like a much better solution, though it could easily be done post-installation by the user.

**Comment by chexo3** [comment]
So it makes sense to allow that feature, because it's simpler than explaining to some 11 year old who just got minecraft how they need to install a different launcher entirely.

**Comment by TwoLeggedCat** [comment]
> how they need to install a different launcher entirely.

Again, why do we keep coming back to MultiMC? This is possible with Mojang's launcher; you would not need to explain this.

**Comment by TwoLeggedCat** [comment]
As it is, mod users create new installations in the vanilla launcher for different versions. This is the best way for this to be done, as it encourages them to use different game directories for different versions. The proposed change would not only make it easier for users to use the same folder across multiple versions, but actually encourage it ("If it were bad, why do they tel me I can do it?"). While I understand the point being made that it may be a bit easier for non-technical users to put mods in folders for the correct version, this is the exact group we would *want* to make use separate installations, as they are the ones who might not know exactly what can lead to world corruption and what cannot. IMO, adding a feature targeted specifically at non-technical users that encourages them to play multiple versions in the same folder is just plain wrong. All they have to do is load one 1.8 world in 1.7 and say "bye bye" to all their chests.

TL;DR: yes, it makes it easier, but being easier makes it easier for people who don't understand what they're doing to mess up.

**Comment by Geolykt** [comment]
While I am a technical person to some degree I wouldn't change the game directory as it results in more clutter, having to download assets twice (okay, that is a minor concern). I also wouldn't use a 3rd party launcher because it too inevitably also causes clutter.

The joining a world with a different version thing is something you cannot avoid (especially because this is only an issue with 1.8 and below, which quilt does not support natively. If any nonstandard quilt implementations should choose to do that they may also choose to fingerprint the saves to do that)
And I guess that using a different game directory in the launcher is *easier* than having to navigate to the .minecraft folder and creating a subdirectory.
I also don't quite understand why using a singular game directory for all profiles should hurt. I've been doing that for multiple years now and never found any issues with it

**Comment by sschr15** [comment]
From experience, this is a (mostly) one line change. Another possibility would be implementing an API for specifying additional folders on the fly.

**Comment by WanderingCat27** [comment]
> > > Can't you use multiple different installations with separate game directories for different versions? We wouldn't want to encourage people to use the same game directory across several versions anyway.
> 
> > 
> 
> > It's complicated and I recall mods not working nicely with that; though that experience must be a few years old by now.
> 
> 
> 
> I have had no complications with such a thing. On Vanilla launcher its as simple as changing the installation folder to somewhere else and then dragging the mods in there instead.

WAIT U CAN DO THAT IN THE VANILLA LAUNCHER.

i wish i knew that its literally the reason i switched to gd (multimc didnt have support for microsoft accs at that time) i feel like this should be the default tho i didnt even know that was an option even as a technical user.

**Comment by teddyxlandlee** [comment]
Different from Forge/FML Mods, Quilt/Fabric Mod can be cross-versions (e.g. 1.16-1.18.2).  
So maybe this could be better:  
```
mods
| - foo-0.1.jar
| - 1.18.2/      # For 1.18.2 only
| | - bar-0.1-1.18.2.jar
| - 1.17.x/      # For 1.17.x, aka >=1.17- <1.18-
| | - bar-0.1-1.18.jar
| - lorem_ipsum_dolor/
| | - mods.properties     # aka mods.txt, to adapt common Windows users
| | - baz-1.0.3.jar
```
`mods.properties`: 
```properties
# optional, default true.
# if false, mods in this directory won't be loaded.
quilt=true
# optional, default *
# blank == *
versions=>=1.16- <1.18-
```

**Comment by TheGlitch76** [comment]
This has been implemented, see https://github.com/QuiltMC/quilt-loader/wiki/Mod-Loading-Behaviour
